File: backend/src/main/java/com/godeltech/mastery/backend/domain/entity/Role.java
package com.godeltech.mastery.backend.domain.entity;

import static javax.persistence.FetchType.EAGER;
import static javax.persistence.GenerationType.IDENTITY;

import com.fasterxml.jackson.annotation.JsonIgnore;
import java.io.Serial;
import java.io.Serializable;
import java.util.Set;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(exclude = "clients")
@Entity
@Table(name = "m_roles")
public class Role implements Serializable {

  @Serial private static final long serialVersionUID = -3884329754279213722L;

  @Id
  @GeneratedValue(strategy = IDENTITY)
  private Long id;

  @Column(name = "role_name")
  private String roleName;

  @ManyToMany(mappedBy = "roles", fetch = EAGER)
  @JsonIgnore
  @ToString.Exclude
  private Set<Client> clients;
}

File: backend/src/main/kotlin/com/godeltech/mastery/backend/domain/entity/Role.kt
package com.godeltech.mastery.backend.domain.entity

import java.io.Serializable
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType.LAZY
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType.IDENTITY
import javax.persistence.Id
import javax.persistence.ManyToMany
import javax.persistence.Table

@Entity
@Table(name = "m_roles")
data class Role(

    @Id
    @GeneratedValue(strategy = IDENTITY)
    val id: Long? = null,

    @Column(name = "role_name")
    var roleName: String? = null,

    @ManyToMany(mappedBy = "roles", fetch = LAZY)
    var clients: List<Client>? = null
) : Serializable
